src/test-util.ts

```diff
@@ -189,39 +199,39 @@ export async function newTab(filename?: string): Promise<Tab> {
   script.type = 'text/javascript';
   script.text = `
     const initialize = async (event) => {
-      let testutil = await import(window.origin + '/src/test-util.ts');
+      let testutil = await import(new URL('/src/test-util.ts', location.href));
       testutil.initializeNewTab(event);
-      window.removeEventListener('message', initialize, false);
+      removeEventListener('message', initialize, false);
     };
-    window.addEventListener("message", initialize, false);
+    addEventListener("message", initialize, false);
   `;
   tab.document.head.appendChild(script);
 
   // Start initialization process, waiting for page to become live.
   const interval = setInterval(() => {
-    tab.postMessage({id: 'init', tabId, path: filename}, window.origin);
+    tab.postMessage({id: 'init', tabId, path}, origin);
   }, 100);
-  await promise;
+  await withTimeout(promise, 3000);
   clearInterval(interval);
 
   return {
     run: async (expr: string) => {
       const id = uuid();
       const {promise, resolve, reject} = resolver<unknown>();
-      tabCallbacks[id] = {resolve, reject};
-      if (expr.search('return ') == -1) {
+      tabCallbacks.set(id, {resolve, reject});
+      if (expr.search('return ') === -1) {
         expr = 'return ' + expr;
       }
-      tab.postMessage({id, expr}, window.origin);
-      const result = await promise;
-      delete callbacks[id];
+      tab.postMessage({id, expr}, origin);
```

Comment:
I actually prefer `window.origin` over `origin`. I hardly every use `window.origin` so seeing it as identifier expression makes me look for the variable.

It is kind of a black art which properties of window are ok to access as identifier expressions so feel free to use your own judgement